agent/s3.go

```diff
@@ -1,41 +1,51 @@
 package agent
 
 import (
-	"errors"
 	"fmt"
+	"net/url"
 	"os"
+	"strings"
 
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/aws/credentials"
 	"github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds"
 	"github.com/aws/aws-sdk-go/aws/endpoints"
 	"github.com/aws/aws-sdk-go/aws/session"
 	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
 	"github.com/buildkite/agent/logger"
 )
 
+const (
+	envS3AccessKeyID             = "BUILDKITE_S3_ACCESS_KEY_ID"
+	envS3SecretAccessKey         = "BUILDKITE_S3_SECRET_ACCESS_KEY"
+	envS3DefaultRegion           = "BUILDKITE_S3_DEFAULT_REGION"
+	envArtifactUploadDestination = "BUILDKITE_ARTIFACT_UPLOAD_DESTINATION"
+)
+
 type credentialsProvider struct {
 	retrieved bool
 }
 
 func (e *credentialsProvider) Retrieve() (creds credentials.Value, err error) {
 	e.retrieved = false
 
-	creds.AccessKeyID = os.Getenv("BUILDKITE_S3_ACCESS_KEY_ID")
-	if creds.AccessKeyID == "" {
-		creds.AccessKeyID = os.Getenv("BUILDKITE_S3_ACCESS_KEY")
+	if v := os.Getenv(envS3AccessKeyID); v != "" {
+		logger.Debug("Found s3 access key id from " + envS3AccessKeyID)
```

Comment:
My brain did struggle to understand those changes! I was like "something looks wrong here, but I can't figure it out" - was hard to understand in the first place :)